### PR TITLE
Add GPU implementation of QR factorization [wip]

### PR DESCRIPTION
**Pull request by nicolov**
## Proposed changes

Add a GPU implementation of QR factorization using the blocked Householder reflection algorithm, see:

Andrew Kerr, Dan Campbell, Mark Richards, QR Decomposition on GPUs
Jan Priessnitz, GPU acceleration of matrix factorization

[Here](https://github.com/ml-explore/mlx/files/14947393/hh.py.zip) is the reference code in numpy for the algorithm.

### Left todo

- [x] clean up handling of batched inputs: slice the inputs/outputs and only pass the slice to the algorithm. Temporaries need only be sized for a single input matrix.
- [ ] share some constants between the kernel and the driver function.
- [x] consider merging the two kernels to compute W.
- [ ] benchmark and optimize grid/block sizes.

## Checklist

Put an `x` in the boxes that apply.

- [ ] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [ ] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/backend/metal/qrf.cpp [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+      }
+
+      const float beta = 2.0f / sq_norm_v;
+      betas_data[jj] = beta;
+    }
+
+    // Apply reflection to the rest of the current block.
+    if constexpr (!RUN_ON_CPU_FOR_DEBUG) {
+      auto kernel = device.get_kernel("qrf_reflect_current_block");
+      auto command_buffer = device.new_command_buffer(stream.index);
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:120]: You don't need to get the command buffer here. Instead get the already open command encoder:

`d.get_command_encoder(stream.index)`

File: mlx/backend/metal/qrf.cpp [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+
+      auto compute_encoder =
+          metal::CommandEncoder(command_buffer->computeCommandEncoder());
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:123]: Remove this, see above

File: mlx/backend/metal/qrf.cpp [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+
+      // Specify the grid dimension and use the largest possible threads groups,
+      // see
+      // https://developer.apple.com/documentation/metal/compute_passes/calculating_threadgroup_and_grid_sizes
+      const MTL::Size threads_per_grid(m, 1, 1);
+      // todo: share threadgroup size with the kernel shared memory buffer.
+      const MTL::Size threads_per_threadgroup(8, 1, 1);
+      compute_encoder->dispatchThreads(
+          threads_per_grid, threads_per_threadgroup);
+      compute_encoder->endEncoding();
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:144]: Don't end the encoding. MLX device will handle this

File: mlx/backend/metal/qrf.cpp [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+    }
+  }
+
+  // Precompute Yp = Y.T @ Y
+  if constexpr (!RUN_ON_CPU_FOR_DEBUG) {
+    auto kernel = device.get_kernel("qrf_compute_yp");
+    auto command_buffer = device.new_command_buffer(stream.index);
+
+    auto compute_encoder =
+        metal::CommandEncoder(command_buffer->computeCommandEncoder());
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:192]: Same comments as above.

File: mlx/backend/metal/qrf.cpp [hunk 5]
```diff
@@ -0,0 +1,696 @@
...
+    const MTL::Size threads_per_grid(R, R, 1);
+    const MTL::Size threads_per_threadgroup(
+        std::min(threads_per_grid.width, kernel->threadExecutionWidth()),
+        std::min(
+            threads_per_grid.height, kernel->maxTotalThreadsPerThreadgroup()),
+        1);
+    compute_encoder->dispatchThreads(threads_per_grid, threads_per_threadgroup);
+    compute_encoder->endEncoding();
+
+    device.commit_command_buffer(stream.index);
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:213]: Don't end encoding or commit here. MXL device will handle it.

**Comment by nicolov** [comment]
@awni I tried to apply your comments and pushed https://github.com/ml-explore/mlx/pull/975/commits/501b889de904adf3b60f4b19d16b3ee46bdc5db2 to avoid creating a new command buffer for each kernel, but I get:

```
-[AGXG13XFamilyCommandBuffer tryCoalescingPreviousComputeCommandEncoderWithConfig:nextEncoderClass:]:1015: failed assertion `A command encoder is already encoding to this command buffer'
```

**Comment by awni** [comment]
Did you manually make a command encoder  from the command buffer? MLX manages an active command encoder so you should not make it directly. Rather call the `device.get_command_encoder()` to get the active encoder.

**Comment by nicolov** [comment]
> Rather call the `device.get_command_encoder()` to get the active encoder.

I also tried doing that in https://github.com/ml-explore/mlx/pull/975/commits/b979ccf3e4051b99320e5d69abe138359c9f0660 which just produces the wrong result.

**Comment by nicolov** [comment]
I also tried tracing and XCode complains about redundant bindings. Should I somehow refactor how I bind buffers to the encoder?

Screenshot 2024-04-15 at 3 09 58 PM

File: mlx/backend/metal/qrf.cpp [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+  // Fill Q with identity.
+  for (int k = 0; k < batch_size; k++) {
+    for (int i = 0; i < m; i++) {
+      for (int j = 0; j < m; j++) {
+        const auto batch_offset = m * n * k;
+        const auto loc = batch_offset + colmajor_idx(i, j, m);
+        q.data<float>()[loc] = i == j ? 1 : 0;
+      }
+    }
+  }
```
Comment (awni) [inline mlx/backend/metal/qrf.cpp:593]: That should probably be a kernel.

File: mlx/utils.cpp [hunk 7]
```diff
@@ -172,7 +172,7 @@ inline size_t elem_to_loc(
 
 template <typename T>
 void print_subarray(std::ostream& os, const array& a, size_t index, int dim) {
-  int num_print = 3;
```
Comment (awni) [inline mlx/utils.cpp:175]: Why did you change that?

Comment (awni) [reply mlx/utils.cpp:175]: Just for debugging?

**Comment by nicolov** [comment]
I fixed the code (needed to introduce one more kernel to ensure the atomics were synchronized properly across different threadgroups). It's a bit slow, so I'll try to improve it now:

```
  device     n  time_ms
0    cpu  2000    99.39
1    gpu  2000   283.36
```

**Comment by awni** [comment]
@nicolov are you planning to come back to this?

**Comment by nicolov** [comment]
I compared with Pytorch, which can use either magma or cusolver. I used a cloud machine with A100 and a desktop with a 1050 (which should be in the ballpark of my M1 Max):

|      | magma | cusolver | cpu  |
|------|-------|----------|------|
| 1050 | 8.8   | 11.7     | 14.1 |
| A100 | 18.2  | 6.3      | 21   |

The magma backend (see [here](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp#L1802)) ends up calling `magma_sgeqrf2_gpu(m, n, dA, ldda, tau, info)` ([here](https://github.com/pytorch/pytorch/blob/93a33bf3ac0b4c9560b49780eabcad2f76dcf43e/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp#L616C5-L616C22)) which is an hybrid CPU/GPU algorithm using queues to overlap CPU compute (factorization of small submatrices) and GPU compute (transformation on the remainder of the matrix).

I'm not sure if it makes sense to implement such a hybrid in mlx, as it would require a lot of synchronization points. @awni what do you think?

Benchmark code:

```
import torch
a = torch.randn((1000, 1000))
a_cuda = a.to('cuda')

%timeit torch.linalg.qr(a, 'complete')

torch.backends.cuda.preferred_linalg_library('magma')
%timeit torch.linalg.qr(a_cuda, 'complete')

torch.backends.cuda.preferred_linalg_library('cusolver')
%timeit torch.linalg.qr(a_cuda, 'complete')
```

**Comment by nicolov** [comment]
I'll close this PR, will re-open if I find an algorithm that's fast enough.
